driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSDownloadPublisherImpl.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2016 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal;
+
+import com.mongodb.client.gridfs.model.GridFSFile;
+import com.mongodb.reactivestreams.client.Success;
+import com.mongodb.reactivestreams.client.gridfs.GridFSDownloadPublisher;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.nio.ByteBuffer;
+
+@SuppressWarnings("deprecation")
+public class GridFSDownloadPublisherImpl implements GridFSDownloadPublisher {
+    private final com.mongodb.reactivestreams.client.gridfs.GridFSDownloadStream gridFSDownloadStream;
+    private int bufferSizeBytes;
+
+    GridFSDownloadPublisherImpl(final com.mongodb.reactivestreams.client.gridfs.GridFSDownloadStream gridFSDownloadStream) {
+        this.gridFSDownloadStream = gridFSDownloadStream;
+    }
+
+    @Override
+    public Publisher<GridFSFile> getGridFSFile() {
+        return gridFSDownloadStream.getGridFSFile();
+    }
+
+    @Override
+    public GridFSDownloadPublisher bufferSizeBytes(final int bufferSizeBytes) {
+        this.bufferSizeBytes = bufferSizeBytes;
+        return this;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super ByteBuffer> s) {
+        s.onSubscribe(new GridFSDownloadSubscription(s));
+    }
+
+    class GridFSDownloadSubscription implements Subscription {
```

Comment:
This is where the complexity lies but it highlights some of the issues consuming the API in a meaningful way.